File: bin/create_ingest_table.pl
#!/usr/bin/env perl
########################################################################
#
#  $Id: arrm 1577 2008-08-14 20:04:19Z dadams $
#
#  $Rev:: 1577                             $:  # Revision of last commit.
#  $LastChangedBy:: dadams                 $:  # Author of last commit. 
#  $LastChangedDate:: 2008-08-14 15:04:19 #$:  # Date of last commit.
#
#  Author: 
#         Darren Adams (dadams@ncsa.uiuc.edu)
#
#  Developed at: 
#  The National Center for Supercomputing Applications (NCSA).
#
#  Copyright (C) 2007 Board of Trustees of the University of Illinois. 
#  All rights reserved.
#
################################################################################
use strict;
use warnings;
use Carp;
use Getopt::Long;
use Pod::Usage;
use FindBin;
use Exception::Class::DBI;
#use lib ("$FindBin::Bin/../lib/perl5","$FindBin::Bin/../lib");
#use DB::DESUtil;
use coreutils::DESUtil;
use DB::IngestUtils;

$| = 1;

my ($run,$table_name,$main_table,$tablespace,$temp_table);

Getopt::Long::GetOptions(
   'run=s' => \$run,
   'table-name=s' => \$table_name,
   'main-table=s' => \$main_table,
   'tablespace=s' => \$tablespace,
   'temp-table-name=s' => \$temp_table,
);

if (!defined($main_table)) {
    Pod::Usage::pod2usage (
        -verbose => 0, -exitval => 2, -output => \*STDERR,
        -message => 'Must provide the name of the main table which will be used as a template for the temp table.'
    );
}

if (!defined($temp_table)) {
    Pod::Usage::pod2usage (
        -verbose => 0, -exitval => 2, -output => \*STDERR,
        -message => 'Must provide the name of the temporary table which will be used to stage the data that is being loaded into the table named after the run.'
    );
}

# Make a new table name
if (!$table_name) {
  if ($run) {
    $table_name = "P$run";
  }
  else {
    Pod::Usage::pod2usage (
    -verbose => 0, -exitval => 2, -output => \*STDERR,
    -message => 'Must provide a run string or table-name to name tmp table'
  );
  }
}

my $db = coreutils::DESUtil->new();

$db->createTableAs({
  'table_name' => $table_name,
  'as' => $main_table,
  'existok' => 1,
  'commit' => 0,
  'tablespace' => $tablespace
});

$db->createTableAs({
  'table_name' => $temp_table,
  'as' => $main_table,
  'existok' => 1,
  'commit' => 0,
  'temp' => 1
}) if not checkIfTableExists($db,$temp_table);

# No longer need index on tmp table
#my $index_name = 'I'.$table_name;
#my $index_sql = "CREATE INDEX $index_name ON $table_name(catalogid) PARALLEL 4 NOLOGGING";
#$db->do($index_sql);

$db->disconnect;

exit 0;

################################################################################
# Documentation
################################################################################

=head1 NAME

=head1 SYNOPSIS

=head1 DESCRIPTION

=head1 OPTIONS

File: bin/create_ingest_table.py
#!/usr/bin/env python

# $Id: create_ingest_table.py 11430 2014-01-17 21:41:16Z tomashek $
# $Rev::                                  $:  # Revision of last commit.
# $LastChangedBy::                        $:  # Author of last commit.
# $LastChangedDate::                      $:  # Date of last commit.

__version__ = "$Rev: 11430 $"

import os
import sys
import argparse
from coreutils import desdbi


if __name__ == '__main__':

    parser = argparse.ArgumentParser(description='Create ingest temp table')
    parser.add_argument('--temptable',action='store')
    parser.add_argument('--targettable',action='store')
    parser.add_argument('--request',action='store')

    args, unknown_args = parser.parse_known_args()
    args = vars(args)

    errors = []
    if args['targettable'] == None:
        errors.append("targettable is required")
    if args['temptable'] == None:
        errors.append("temptable is required")
    if args['request'] == None:
        errors.append("request is required")
    if len(errors) > 0:
        sys.stderr.write("ERROR: " + "; ".join(errors) + "\n")
        exit(1)

    tablespace = "DESSE_REQNUM%07d_T" % int(args['request'])

    dbh = desdbi.DesDbi()
    cursor = dbh.cursor()
    print "Creating tablespace %s ..." % tablespace
    cursor.callproc("des_admin.createObjectsTablespace",[tablespace])

    print "Creating temp table %s using %s as a template..." % (args['temptable'], args['targettable'])
    cursor.execute("create table %s tablespace %s as select * from %s where 1=0" % (args['temptable'],tablespace,args['targettable']))
    cursor.close()
    print "Temp table %s created successfully" % args['temptable']

